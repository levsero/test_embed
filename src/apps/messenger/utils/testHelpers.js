import React from 'react'
import { IdManager } from '@zendeskgarden/react-selection'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
import { ThemeProvider, DEFAULT_THEME } from '@zendeskgarden/react-theming'
import createStore from 'src/apps/messenger/store'
import configureMockStore from 'redux-mock-store'

const mockStore = configureMockStore()

export const render = (ui, { render, store, state, themeProps = {} } = {}) => {
  IdManager.setIdCounter(0)
  if (state) {
    store = mockStore(state)
  }
  const reduxStore = store || createStore(state)

  const renderFn = render || rtlRender
  return {
    ...renderFn(
      <Provider store={reduxStore}>
        <ThemeProvider theme={{ ...DEFAULT_THEME, ...themeProps }}>{ui}</ThemeProvider>
      </Provider>
    )
  }
}
