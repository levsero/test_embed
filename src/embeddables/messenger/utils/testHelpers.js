import { IdManager } from '@zendeskgarden/react-selection'
import { render as rtlRender } from '@testing-library/react'

export const render = (ui, { render } = {}) => {
  IdManager.setIdCounter(0)
  const renderFn = render || rtlRender
  return {
    ...renderFn(ui)
  }
}
