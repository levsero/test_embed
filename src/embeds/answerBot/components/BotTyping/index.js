import React from 'react'

import { Container, Loader } from './styles'
import { TEST_IDS } from 'constants/shared'

const BotTyping = () => (
  <Container data-testid={TEST_IDS.AB_TYPING_INDICATOR}>
    <Loader useUserColor={false} />
  </Container>
)

export default BotTyping
