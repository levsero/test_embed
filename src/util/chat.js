import { AGENT_BOT } from 'constants/chat';
import { win, getZendeskHost } from 'utility/globals';
import _ from 'lodash';

/**
 * Return a new array with consecutive numbers combined into an array containing only the start and end (inclusive) number.
 * @param {NumberLike[]} numbers
 * @returns {NumberRange[]} numberRanges
 */
function combineNumbers(numbers) {
  const result = [];
  let last, start;

  const dump = () => start = (result.push(start ? [start, last] : last), undefined);

  numbers.forEach(num => {
    num = parseInt(num, 10);
    if (last === num-1) {
      if (!start) start = last;
    }
    else if (last) dump();
    last = num;
  });

  dump();
  return result;
}

const periodsToString = periods => periods.map(period => `${period.start},${period.end}`).toString();
const stringToPeriods = str => str
  ? _.chunk(str.split(','), 2).map(arr => ({ start: parseInt(arr[0], 10), end: parseInt(arr[1], 10) }))
  : [];

/**
 * Return a FormattedSchedule array that is more amendable to the Widget's rendering needs.
 * @param {Schedule} schedule
 * @return {FormattedSchedule[]}
 */
function formatSchedule(schedule) {
  schedule = { ...schedule, 7: schedule[0] }; // use 7 for Sunday
  delete schedule[0];

  return _.chain(schedule)
    .mapValues(periodsToString)
    .invertBy()
    .map((days, periods) => ({ periods: stringToPeriods(periods), days: combineNumbers(days) }))
    .sortBy(['days'])
    .value();
}

function isAgent(nick) {
  return nick.indexOf('agent:') > -1 && nick !== AGENT_BOT;
}

function isDefaultNickname(name) {
  const nameRegex = new RegExp(/^Visitor [0-9]{3,}$/);

  return nameRegex.test(name);
}

function createChatPopoutWindow(chatPopoutSettings, machineId, locale) {
  const hostName = win.location.hostname;
  let url;

  if (__DEV__) {
    url = '/liveChat.html';
  } else if (hostName.indexOf('-staging.') !== -1) {
    url = 'https://static-staging.zdassets.com/web_widget/latest/liveChat.html';
  } else {
    url ='https://static.zdassets.com/web_widget/latest/liveChat.html';
  }

  url += generateQueryString(chatPopoutSettings, machineId, locale);

  win.open(url, 'Web Widget LiveChat', 'height=600,width=400');
}

function generateQueryString(chatPopoutSettings, machineId, locale) {
  const subdomain = getZendeskHost(document);
  const settings = win.btoa(JSON.stringify(chatPopoutSettings));

  return `?key=${subdomain}&settings=${settings}&mid=${machineId}&locale=${locale}`;
}

export {
  combineNumbers,
  formatSchedule,
  isDefaultNickname,
  isAgent,
  createChatPopoutWindow
};

/**
 * A number, or a string containing a number.
 * @typedef {(number|string)} NumberLike
 */

/**
 * A number representing a single number, or a 2-element array representing a range of consecutive numbers.
 * @typedef {(number|[number, number])} NumberRange
 */

/**
 * An object representing an operating period. Values refer to the number of minutes since midnight
 * (0 means midnight, 1439 means 23:59)
 * @typedef {object} Period
 * @property {number} start
 * @property {number} end
 */

/**
 * An object containing an operating hours schedule. 0 is for Sunday, 1 for Monday, etc.
 * The properties do NOT actually have underscore prefixed.
 * @typedef {object} Schedule
 * @property {Period[]} _0
 * @property {Period[]} _1
 * @property {Period[]} _2
 * @property {Period[]} _3
 * @property {Period[]} _4
 * @property {Period[]} _5
 * @property {Period[]} _6
 */

/**
 * An object containing operating periods and the applicable days. Sunday is represented by 7 here.
 * @typedef {object} FormattedSchedule
 * @property {NumberRange[]} days
 * @property {Periods[]} periods
 */
