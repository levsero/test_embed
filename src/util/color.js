import Color from 'color';

import { settings } from 'service/settings';

const defaultColor = '#78A300';

// Color manipulation and checks using color library
const darkenAndMixColor = (mixAmount, darkenAmount, mixColor = 'gray') => (color) => {
  return color.mix(Color(mixColor), mixAmount).darken(darkenAmount).rgbString();
};
const darkenColor = (amount) => (color) => color.darken(amount).rgbString();
const lightenColor = (amount) => (color) => color.darken(amount).rgbString();
const getLuminosity = (amount) => (color) => color.luminosity() > amount;

// Color checks
const isColorLight = (colorStr) => {
  const color = Color(colorStr);
  // YIQ equation from http://24ways.org/2010/calculating-color-contrast
  const rgb = color.rgb();
  const yiq = (rgb.r * 299 + rgb.g * 587 + rgb.b * 114) / 1000;

  return yiq > 178;
};
const getLightOrDark = (colorStr, light, dark, isLight = isColorLight(colorStr)) => {
  const color = Color(colorStr);

  return isLight ? dark(color) : light(color);
};

// Color calculations
const buttonTextColor = (color) => getLightOrDark(color, () => 'white', darkenAndMixColor(0.3, 0.5));
const listColor = (color) => getLightOrDark(color, () => color, darkenAndMixColor(0.2, 0.4));
const highlightColor = (color) => getLightOrDark(color, lightenColor(0.15), darkenColor(0.1), getLuminosity(0.15));
const constrastColor = (color) => getLightOrDark(color, () => 'white', () => 'black', getLuminosity(0.65));
const border = (color) => getLightOrDark(color, () => 'none', (c) => `1px solid ${darkenAndMixColor(0.2, 0.2)(c)}`);

function generateUserCSS(color = defaultColor) {
  if (validSettingsColor()) {
    color = validSettingsColor();
  }

  const listColorStr = listColor(color);
  const listHighlightColorStr = highlightColor(listColor);
  const buttonTextColorStr = buttonTextColor(color);

  return (`
    .rf-CheckboxGroup__checkbox:checked + span:before,
    .u-userTextColor:not([disabled]) {
      color: ${listColorStr} !important;
      fill: ${listColorStr} !important;
    }
    .u-userTextColor:not([disabled]):hover,
    .u-userTextColor:not([disabled]):active,
    .u-userTextColor:not([disabled]):focus {
      color: ${listHighlightColorStr} !important;
      fill: ${listHighlightColorStr} !important;
    }
    .u-userFillColor:not([disabled]) svg {
      fill: ${color} !important;
    }
    .u-userFillColor:not([disabled]) svg path {
      fill: ${color} !important;
    }
    .u-userBackgroundColor:not([disabled]) {
      background-color: ${color} !important;
      color: ${buttonTextColorStr} !important;
      fill: ${buttonTextColorStr} !important;
      border: ${border(color)} !important;
      svg {
        color: ${buttonTextColorStr} !important;
        fill: ${buttonTextColorStr} !important;
      }
    }
    .u-userBackgroundColor:not([disabled]):hover,
    .u-userBackgroundColor:not([disabled]):active,
    .u-userBackgroundColor:not([disabled]):focus {
      background-color: ${highlightColor(color)} !important;
    }
    .u-userBorderColor:not([disabled]) {
      color: ${buttonTextColorStr} !important;
      background-color: transparent !important;
      border-color: ${buttonTextColorStr} !important;
    }
    .u-userBorderColor:not([disabled]):hover,
    .u-userBorderColor:not([disabled]):active,
    .u-userBorderColor:not([disabled]):focus {
      color: ${constrastColor(color)} !important;
      background-color: ${listHighlightColorStr} !important;
      border-color: ${listHighlightColorStr} !important;
    }
    .u-userLinkColor a {
      color: ${color} !important;
    }
    .u-userStrokeColor {
      stroke: ${color} !important;
    }
    .u-userHeaderColor {
      background: ${color} !important;
      color: ${buttonTextColorStr} !important;
    }
  `);
}

function generateWebWidgetPreviewCSS(color) {
  if (validSettingsColor()) {
    color = validSettingsColor();
  }

  return (`
    .u-userBackgroundColor:not([disabled]) {
      background-color: ${color} !important;
    }
    .u-userBackgroundColor:not([disabled]):hover,
    .u-userBackgroundColor:not([disabled]):active,
    .u-userBackgroundColor:not([disabled]):focus {
      background-color: ${highlightColor(color)} !important;
    }
  `);
}

function validSettingsColor() {
  const settingsColor = settings.get('color.theme');
  const settingsColorValid = settingsColor &&
                             isValidHex(settingsColor);

  return settingsColorValid ? settingsColor : null;
}

function isValidHex(color) {
  const validRegex = /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i;

  return validRegex.test(color);
}

export {
  generateUserCSS,
  generateWebWidgetPreviewCSS,
  validSettingsColor
};
