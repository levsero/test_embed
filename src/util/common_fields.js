import { Field } from 'component/field/Field';
import React from 'react';

export function renderTextField(name, label, value, required = false) {
  return (
    <Field
      label={label}
      required={required}
      value={value}
      name={name} />
  );
}

export function renderTextAreaField(name, label, value, required = false) {
  return (
    <Field
      label={label}
      required={required}
      value={value}
      input={<textarea rows='3' />}
      name={name} />
  );
}

export function renderPhoneField(name, label, value, required = false) {
  return (
    <Field
      label={label}
      required={required}
      value={value}
      name={name} />
  );
}

export function renderEmailField(name, label, value, required = false) {
  return (
    <Field
      label={label}
      required={required}
      value={value}
      pattern="[a-zA-Z0-9!#$%&'*+/=?^_`{|}~\-`']+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~\-`']+)*@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?" // eslint-disable-line
      name={name} />
  );
}
