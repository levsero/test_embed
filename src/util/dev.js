if (__DEV__) {
  require('react-hot-loader')
}
