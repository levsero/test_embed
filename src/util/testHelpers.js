import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { GET_ACCOUNT_SETTINGS_REQUEST_SUCCESS } from 'src/redux/modules/chat/chat-action-types';
import { i18n } from 'src/service/i18n';

export const dispatchChatAccountSettings = (store, settings) => {
  store.dispatch({
    type: GET_ACCOUNT_SETTINGS_REQUEST_SUCCESS,
    payload: settings
  });
};

export const clearDOM = () => {
  document.getElementsByTagName('html')[0].innerHTML = '';
};

export const noopReactComponent = () => class extends Component {
  static propTypes = {
    className: PropTypes.string,
    children: PropTypes.node,
  };

  render() {
    return (
      <div className={this.props.className}>
        {this.props.children}
      </div>
    );
  }
};

/*
  For testing reducers (duh).
  Pass it your reducer and an array of parameter objects.
  - If parameter object is a flat action payload, the test will use
  snapshot testing. The action can also be explicitly specified using
  an `action` key.
  - If parameter object contains key `initialState`, the value
  of that key will be used as initial state for the reducer.
  - If parameter object contains key `expected`, the value
  of that key will be used to assert against the result of the reducer.
*/
export const testReducer = (reducer, actions) => {
  actions.forEach((params) => {
    const { expected, initialState, extraDesc } = params;
    const action = params.action || params;
    const basicTestDesc = `${reducer.name}, action: ${action.type}`;
    const testDesc = extraDesc ? `${basicTestDesc} ${extraDesc}` : basicTestDesc;

    test(testDesc, () => {
      const reduced = reducer(initialState, action);

      if ('expected' in params) {
        expect(reduced).toEqual(expected);
      } else {
        expect(reduced).toMatchSnapshot();
      }
    });
  });
};

/*
  For testing basic translation selectors made using reselect's `createSelector`.
  And by basic, I mean of the i18n.getSettingTranslations(setting) || i18n.t('default.translation') kind.
  The resultFunc needs to have an arity of 2 for the unused locale param.
*/
export const testTranslationStringSelector = (selector) => {
  const subject = selector.resultFunc;
  const defaultTranslation = 'default translation';
  const settingsTranslation = 'settings translation';

  if (!subject || subject.length !== 2) {
    throw new Error('For testing create selector resultFuncs with arity of 2');
  }

  describe(selector.name, () => {
    test.each([
      [undefined,            null,  defaultTranslation ],
      [settingsTranslation,  null,  settingsTranslation]
    ])('resultFunc(%p, %p) returns %p',
      (settingString, locale, expected) => {
        jest.spyOn(i18n, 't').mockReturnValue(defaultTranslation);
        jest.spyOn(i18n, 'getSettingTranslation').mockReturnValue(settingString);

        expect(subject(settingString, locale)).toEqual(expected);

        i18n.t.mockRestore();
        i18n.getSettingTranslation.mockRestore();
      }
    );
  });
};
