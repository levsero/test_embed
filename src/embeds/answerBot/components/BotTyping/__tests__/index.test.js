import { render } from 'src/util/testHelpers'
import React from 'react'

import BotTyping from '../index'

test('renders the expected classes', () => {
  const { container } = render(<BotTyping />)

  expect(container).toMatchSnapshot()
})
