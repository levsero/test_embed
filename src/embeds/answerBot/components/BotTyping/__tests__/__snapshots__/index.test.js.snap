// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the expected classes 1`] = `
.c0 {
  display: inline-block;
  max-width: 100%;
  float: left;
  background: #f3f3f3;
  border-radius: 1.1428571428571428rem;
  padding: 0.5714285714285714rem 0.8571428571428571rem;
  margin-left: 2.857142857142857rem;
}

.c1 {
  margin: 0.2857142857142857rem;
}

.c1 div {
  width: 0.39285714285714285rem;
  height: 0.39285714285714285rem;
  background-color: #68737d;
  margin-right: 0.2857142857142857rem;
}

<div>
  <div>
    <div
      class="c0"
      data-testid="answer-bot-typing-indicator"
    >
      <div
        aria-busy="true"
        aria-live="polite"
        class="container c1"
        data-testid="Icon--ellipsis"
      >
        <div
          class="
      circle 
      bounce
      
    "
        />
        <div
          class="
      circle 
      bounce
      
    "
        />
        <div
          class="
      circle 
      bounce
      
    "
        />
      </div>
    </div>
  </div>
</div>
`;
